### PR TITLE
Cloud storage initial implementation, with local folder provider.

### PR DESCRIPTION
**Pull request by njordan64**
# Phase 1 of Cloud Storage Support

## Description
Beginning of cloud-storage support. This is the initial work. It contain logic for syncing game states with a provider. The only provider available is **local_folder**, which just syncs the game states folder with a local folder. By default, cloud storage support is not built in.

## Building
When running configure, need to add the argument `--enable-cloud_storage_debug`

## Configuration
No configuration is needed.

## Local Folder Provider
If the environment variable `RETRO_CLOUD_STORAGE_TEST` is set, then it will be used as the folder to sync to. Otherwise, it will try to create the folder `retro_cloud_storage_test` in the tmp folder.

## Reviewing
I would suggest going in the following order:
1. Look over the code for the local folder provider
2. Look over the logic in `cloud-storage.c` for how it syncs files
3. Look over `hex.c` (quite simpe in comparison)
4. Look over the changes for adding hooks to call cloud storage
5. Look over the configure script changes

### CODE REVIEW

**Comment by jdgleaver** [comment]
@njordan64 Thank you for this! It is far more manageable :)

I can see a few very trivial issues, so as a first pass I'll go through everything a prepare a diff that you can apply on-top of the PR (easier than posting individual comments!). I'll then do all the ASAN tests, etc., and we'll get another pair of eyes to double check the logic + API. I would appreciate it if you didn't push any more commits to the branch until we're done with this, just to avoid any overlapped effort.

We're kinda busy with the new build infrastructure at the moment (pre-Christmas crunch to get everything done...), so it may take a little while to get this reviewed - but I don't anticipate any blocking issues this time :)

Please also keep the other PR open for now, as it may be useful for reference.

**Comment by njordan64** [comment]
@jdgleaver Take your time. I'll leave this branch alone for now.

**Comment by m4xw** [comment]
I will try to get a review and a list of suggestions done between the holidays.

**Comment by njordan64** [comment]
> I will try to get a review and a list of suggestions done between the holidays.

@m4xw Thanks for trying to make time to look this over. The more eyes on it the better.

**Comment by njordan64** [comment]
@jdgleaver @m4xw Just checking in, since this may have been lost over the holidays.

**Comment by m4xw** [comment]
We are actually still busy with new infra and real life :/

**Comment by jdgleaver** [comment]
@njordan64 Please accept my apologies. I had no idea going into it how much work would be involved with the new build infrastructure (ended up working through the entire 'holidays' - the whole experience has been very unpleasant). We are indeed still busy at the moment, but this PR is very much on my TODO list...

**Comment by njordan64** [comment]
> @njordan64 Please accept my apologies. I had no idea going into it how much work would be involved with the new build infrastructure (ended up working through the entire 'holidays' - the whole experience has been very unpleasant). We are indeed still busy at the moment, but this PR is very much on my TODO list...

@jdgleaver Is there any way that I could help with the build infrastructure?

**Comment by jdgleaver** [comment]
@njordan64 That's really kind of you to offer :)

Unfortunately, in part due to the earlier hacking incident that forced the switchover, much of the work is being done around private servers/project groups - so the issues that are holding us up right now aren't really things that people outside the main team (in as much as we have a team...) can readily access or contribute to.

Hopefully it won't take too much longer, though. I, for one, am very much looking forward to getting back to regular dev work...

**Comment by m4xw** [comment]
We didnt forget :)

**Comment by twinaphex** [comment]
Yeah, since the key signing is taking way too long over at our end with no real end in sight, let us start focusing on getting this fast tracked. 

@njordan64 How  do you think we should proceed?

**Comment by twinaphex** [comment]
OK, I wasn't aware @jdgleaver and @m4xw were already dealing with this already. I'm very grateful they are though. Let's indeed move on as planned.

**Comment by njordan64** [comment]
For getting started, first steps would be:
* Test this out. I did some testing, but it needs to be tested by others.
* I can rebase this to bring it up to date
* I am normally a Java programmer, so there could be some issues with my C code. Let's start with the most obvious and repetive errors. Comment those and I'll update.
* Then get into the structure and logic. This will take more time.

**Comment by njordan64** [comment]
Closing this merge request. It was opened almost 4 months ago. There is no traction on this and no sign of that changing. If there is interest and a path forward in the future, let me know and I'll reopen this.

I'll rebase the original cloud-storage branch in case anyone wants to use it.

**Comment by twinaphex** [comment]
Hi there @njordan64 ,

please don't give up on us. We have been having a really hard time with the server rebasing and we finally got out of the woods on that. @jdgleaver is definitely meaning to look into this but he is on easter break for now. He said that he would definitely look into this when given the chance but he didn't want to post here again only for there to be no progress so he wanted to post here when he finally had some progress to report.

Please don't give up on this, we want this incorporated as badly as you do.

Maybe a more efficient way to get stuff done also would be to meet me on Discord. You can visit us on our Discord channel, I am known as Autechre there. You can send me a DM there and from there I can add you to developer groups where we can maybe get some momentum going on this, get more developers involved in testing, etc.

**Comment by twinaphex** [comment]
This is our Discord channel - https://discordapp.com/invite/27Xxm2h

**Comment by twinaphex** [comment]
Hi @njordan64 , I've asked @davidgfnet and he definitely wants to look into this now with the express purpose of being able to turn this into something that can be merged. Would you be willing to give us another chance given this new plan of action?

**Comment by m4xw** [comment]
Funnily enough just thought of this PR earlier, but I am dying in irl work currently...
Didn't want to bump it since my last reminder was still fairly recent.

**Comment by njordan64** [comment]
@twinaphex We probably should talk about this on Discord. Please don't feel pressured to move this forward. I trust that this is given appropriate priority within the libretro project.

Given that this is a massive amount of code from an outside developer, it probably only makes sense as a fork. If many people used this feature one day, then a core developer could reimplement this.

**Comment by njordan64** [comment]
I have no plans of deleting this branch, so anyone could still add comments here, although it is out of date.

**Comment by twinaphex** [comment]
Nah, we definitely want it in master. A separate fork is not needed for this. 

Definitely meet me up on Discord and I'll get you in touch with davidgfnet and hopefully we get somewhere fast with this. I'm not proud myself of the 4 months it's taken with this PR so far but to our defense, all of us underestimated the scope and time it would take to get the buildbot all done. Under different circumstances this would all have been wrapped up already. Anyway, I'm confident that we can still get this done fairly soon with David looking at it together with you.

**Comment by njordan64** [comment]
Reopening for review.

File: cloud-storage/cloud_storage.c [hunk 1]
```diff
@@ -0,0 +1,909 @@
...
+
+static void _operation_thread_loop(void *user_data)
+{
+   while (!_shutdown)
+   {
+      struct _operation_queue_item *task;
+
+      slock_lock(_operation_queue_mutex);
+
+      scond_wait_timeout(_operation_condition, _operation_queue_mutex, 500);
```
Comment (davidgfnet) [inline cloud-storage/cloud_storage.c:166]: Why do you wake up the condition periodically? You signal it from the exit code, so it will shutdown gracefully and immediately. Right?

File: cloud-storage/cloud_storage.c [hunk 2]
```diff
@@ -0,0 +1,909 @@
...
+      if (_operation_queue)
+      {
+         _operation_queue = NULL;
+      }
+
+      slock_unlock(_operation_queue_mutex);
+
+      while (task)
+      {
+         struct _operation_queue_item *next_task;
```
Comment (davidgfnet) [inline cloud-storage/cloud_storage.c:183]: I don't think this queue code is properly implemented, but I dont have more proof than reading it.
In the queue add item routine, it modifies the last element next field to queue a new operation. Here, the task next is accessed without grabbing the lock, that sounds like trouble to me.
I'd recommend you to get rid of this while loop and simply "pop" the front element before (in the area where you grab the lock). That way _shutdown is also evaluated, allowing to early exit the queue if needed.

File: cloud-storage/hex.c [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+   '9',
+   'a',
+   'b',
+   'c',
+   'd',
+   'e',
+   'f'
+};
+
+char *bytes_to_hex_str(unsigned char *bytes, uint64_t len)
```
Comment (davidgfnet) [inline cloud-storage/hex.c:46]: Wanna move this to libretro-common/string/stdstring.c (if it's not there already)? I think this might already exist, but I'm quite new to RA, so I'm just assuming

File: cloud-storage/cloud_storage.h [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+
+void cloud_storage_shutdown(void);
+
+void cloud_storage_sync_files(void);
+
+void cloud_storage_upload_file(folder_type_t folder_type, char *file_name);
+
+void cloud_storage_item_free(cloud_storage_item_t *item);
+
+char *cloud_storage_get_md5_hash(char *absolute_filename);
```
Comment (davidgfnet) [inline cloud-storage/cloud_storage.h:134]: Does this need to be exposed here? I assume nothing in retroarch.c (and common files) need to use this right?

**Review by davidgfnet — CHANGES_REQUESTED** [review]

File: cloud-storage/cloud_storage.c (hunk 2 above)
Comment (njordan64) [reply cloud-storage/cloud_storage.c:183]: On line 164, I get the lock. I then assign the variable **task** on line 173 to be the operation queue and set the operation queue to NULL. Finally the lock is released on line 179.

The basic idea is to take all tasks from the queue and then process them.

This ties in to your comment above. Since it is checking for tasks every 0.5 seconds, I am trying to take all tasks. If I can get the condition working as expected, then this can be changed to only work on one task.

File: cloud-storage/cloud_storage.c (hunk 1 above)
Comment (njordan64) [reply cloud-storage/cloud_storage.c:166]: I ran into an issue with the condition variable.
1. Add sync task to operation queue.
2. Signal the condition variable
3. The operation thread starts up and runs this function
4. Fails to wakeup when waiting on the condition variable, since the signal was already sent.

Thinking about this now, I should just skip the condtion wait if there are items in the queue.

File: cloud-storage/cloud_storage.h (hunk 4 above)
Comment (njordan64) [reply cloud-storage/cloud_storage.h:134]: This does not need to be exposed here. It is only used within cloud storage. I'll move it.

File: cloud-storage/hex.c (hunk 3 above)
Comment (njordan64) [reply cloud-storage/hex.c:46]: I'll take a look and remove this function from here.

File: cloud-storage/providers/local_folder/get_item_metadata.c [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+cloud_storage_item_t *cloud_storage_local_folder_get_file_metadata(cloud_storage_item_t *file)
+{
+   cloud_storage_item_t *metadata;
+
+   if (!path_is_valid(file->id)) {
+      return NULL;
+   }
+
+   metadata = (cloud_storage_item_t *)calloc(sizeof(cloud_storage_item_t), 1);
+   metadata->id = (char *)malloc(strlen(file->id) + 1);
```
[the remark this answers is not in the document]
Comment (njordan64) [reply cloud-storage/providers/local_folder/get_item_metadata.c]: Thanks, these have been replaced.

File: cloud-storage/cloud_storage.c (hunk 1 above)
Comment (njordan64) [reply cloud-storage/cloud_storage.c]: This has been improved as I mentioned above. Now it will only wait on the condition variable if the operation queue is empty. Also changed to only pull off the first item from the queue.

File: cloud-storage/cloud_storage.c (hunk 2 above)
Comment (njordan64) [reply cloud-storage/cloud_storage.c]: This has been changed to only work on the first item in the queue (which is the oldest item).

File: cloud-storage/hex.c (hunk 3 above)
Comment (njordan64) [reply cloud-storage/hex.c]: Moved to stdstring.

File: cloud-storage/cloud_storage.h (hunk 4 above)
Comment (njordan64) [reply cloud-storage/cloud_storage.h]: Moved to a new header file that is only used by the providers.

**Comment by njordan64** [comment]
Updated to depend on threads being enabled. Should fix a compilation issue when threads are disabled.

File: cloud-storage/providers/local_folder/get_item_metadata.c [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+cloud_storage_item_t *cloud_storage_local_folder_get_file_metadata_by_name(
+   cloud_storage_item_t *folder,
+   char *filename)
+{
+   char *absolute_path;
+   const char *base_dir;
+   size_t absolute_path_len;
+   cloud_storage_item_t *metadata;
+   base_dir = cloud_storage_local_folder_get_base_dir();
+   absolute_path_len = strlen(base_dir) + strlen(filename) + 4;
```
[the remark this answers is not in the document]
Comment (njordan64) [reply cloud-storage/providers/local_folder/get_item_metadata.c:86]: I don't see how the functions **fill_pathname_join_delim_concat** or **fill_pathname_join_delim** help here. I am already using **fill_pathname_join**. For any of those I need a pre-allocated output variable and its length. I could simplify a little by using a hard-coded value for absolute_path_len (perhaps 4k).

I could create a new function the adds a filename to a path and handles the memory allocation. The would be less code here.

File: cloud-storage/cloud_storage.c [hunk 7]
```diff
@@ -0,0 +1,887 @@
...
+               _sync_files_execute(task->operation_parameters.sync.folder_type);
+               break;
+            case UPLOAD_FILE:
+               _upload_file_execute(task->operation_parameters.upload.folder_type, task->operation_parameters.upload.local_file);
+               break;
+            case DOWNLOAD_FILE:
+               break;
+         }
+
+         _free_operation_queue(task);
```
Comment (davidgfnet) [inline cloud-storage/cloud_storage.c:201]: Is this ok or should it be _free_operation_queue_item ?

File: cloud-storage/providers/local_folder/get_item_metadata.c (hunk 6 above)
[the remark this answers is not in the document]
Comment (davidgfnet) [reply cloud-storage/providers/local_folder/get_item_metadata.c:86]: Dont worry too much then. Perhaps might make sense to contribute some path manipulation functions to the common code. Do as you see fit!

**Review by davidgfnet — APPROVED** [review]
I left two more comments, but other than that it looks good to me.
Two things I'd like to get improved in future PRs would be:
 - Make upload/download interruptible somehow. Perhaps using some progress callbacks? I think it could be a good solution to allow aborting some long operations and at the same time having a mechanism to give the users some UI-feedback on the progress. 
 - You do some queue optimizations such as removing queued operations that have been superseeded by other operations. This is mostly correct but I've seen some that might not be (like when it completes an upload it checks for pending items that upload the same file). Perhaps having some sort of timestamp would help? Or is this already covered by the hashing mechanism?

Awesome work! Very happy to get this in RA!

File: cloud-storage/cloud_storage.c (hunk 7 above)
Comment (njordan64) [reply cloud-storage/cloud_storage.c]: This worked since **task->next** was set to NULL earlier. The code for **free_operation_queue** and **free_operation_queue_item** needed to be cleaned up though.

Went with **free_operation_queue_item** here since it is just one item.

**Comment by njordan64** [comment]
@davidgfnet Thanks for the feedback. I'll work on your suggestions for the next MR. I might also try to add support for on the fly reordering of the sync operations. This would help during startup. If the user loads a game during sync, it could move the downloading of states/saves to be as soon as possible to minimize the delay until the player can play.

Aborting would be useful. I'm planning to provide UI feedback using the task queue. How could the user abort the task? Is there a better way to provide UI feedback?

**Comment by davidgfnet** [comment]
Hey! I'd appreciate that you test the thing a bit before pushing. It's easy to get carried away and do a last minute change and pushing it directly ^^'
TLDR it crashes, some info for easy debugging:

#3  0x0000000000ba22fb in cloud_storage_local_folder_download_file (file_to_download=0x7fffb400b560, local_file=0x7fffb43274d0 "/home/hellothere/.config/retroarch/states/retroarch.cfg")
    at cloud-storage/providers/local_folder/download_file.c:41
41	   src_file = filestream_open(file_to_download->id, RETRO_VFS_FILE_ACCESS_READ, RETRO_VFS_FILE_ACCESS_HINT_NONE);
(gdb) list
36	   RFILE *src_file;
37	   RFILE *dest_file;
38	   uint8_t buffer[BUFFER_SIZE];
39	   uint8_t bytes_read;
40	
41	   src_file = filestream_open(file_to_download->id, RETRO_VFS_FILE_ACCESS_READ, RETRO_VFS_FILE_ACCESS_HINT_NONE);
42	   if (!src_file)
43	   {
44	      return false;
45	   }
(gdb) p file_to_download->id
$4 = 0x0
(gdb) p file_to_download
$5 = (cloud_storage_item_t *) 0x7fffb400b560
(gdb) p *file_to_download
$6 = {id = 0x0, name = 0x7fffb400b3d0 "retroarch.cfg", last_sync_time = 1618430864, item_type = CLOUD_STORAGE_FILE, type_data = {file = {hash_type = CLOUD_STORAGE_HASH_MD5, 
      hash_value = 0x7fffb400b6f0 "582adb3371c4ec55e3d053db08dcd97c", download_url = 0x0}, folder = {children = 0x0}}, next = 0x7fffb400b6a0}
(gdb)

**Comment by njordan64** [comment]
I was able to get a segfault from a similar test. It looks like this is down to how I am parsing the value for the save states.

I need some clarification on how each of the file should be handled.

1. /foo/bar/.state -- From what I can see, this is the base directory (up to and including the '/') for saving game states. From the '.' onwards is the extension to use for game states.
2. /foo/bar/baz.state -- Will this mean that all save states go to the file /foo/bar/game.state?
3. /foo/bar/baz -- Should I consider this the directory where game states are saved?
4. foo -- Not an absolute path, is this considered a directory relative to the current directory?
5. foo.state -- Is this similar to case 2, but as a file in the current directory?

Once this is clarified, I can go through and create a function to properly handle these cases.

File: cloud-storage/cloud_storage.c [hunk 8]
```diff
@@ -0,0 +1,888 @@
...
+
+   bytes_read = filestream_read(file, buffer, BUFFER_SIZE);
+   while (bytes_read > 0)
+   {
+      MD5_Update(&md5_ctx, buffer, bytes_read);
+      bytes_read = filestream_read(file, buffer, BUFFER_SIZE);
+   }
+
+   filestream_close(file);
+   checksum = (unsigned char *)calloc(16, sizeof(char));
```
Comment (davidgfnet) [inline cloud-storage/cloud_storage.c:280]: Just allocate any small const buffers in the stack :)

File: cloud-storage/cloud_storage.c [hunk 9]
```diff
@@ -0,0 +1,888 @@
...
+         return;
+   }
+
+   if (!remote_folder || strlen(raw_dir) == 0)
+   {
+      return;
+   }
+
+   dir_name = strdup(raw_dir);
+   if (strrchr(dir_name, '.') > strrchr(dir_name, '/'))
```
Comment (davidgfnet) [inline cloud-storage/cloud_storage.c:523]: The code has literally 0 comments. This is one of those places where one comment would save me some time :)

File: cloud-storage/cloud_storage.c [hunk 10]
```diff
@@ -0,0 +1,888 @@
...
+      if (queue_item->operation == UPLOAD_FILE || queue_item->operation == DOWNLOAD_FILE)
+      {
+         if (!previous_queue_item)
+         {
+            _free_operation_queue_item(queue_item);
+         } else
+         {
+            struct _operation_queue_item *next_item;
+
+            next_item = queue_item->next;
```
Comment (davidgfnet) [inline cloud-storage/cloud_storage.c:792]: Here you are deleting the current item (queue_item) and at the end of the block, queue item points to the next item to the item you just deleted.
However the loop will execute `queue_item = queue_item->next` when processing the next iteration, therefore skipping an item in the queue to process. I'm pretty sure this is not on purpose right?
If you want to filter elements of a queue, it's much simpler to builld a new queue with the elements that you do *not* want to delete.
I would add that perhaps it would be a good idea to use a generic queue (not sure whether there's one in RA, but maybe we could create one! With tests :P) and call some generic function like delete_queue_items_filter(queue, evaluate_elem_delete_callback). Just to give you an idea.

File: cloud-storage/cloud_storage.c [hunk 11]
```diff
@@ -0,0 +1,888 @@
...
+      return;
+   }
+
+   for (queue_item = _operation_queue; queue_item != NULL; queue_item = queue_item->next)
+   {
+      if (queue_item->operation == UPLOAD_FILE || queue_item->operation == DOWNLOAD_FILE)
+      {
+         if (!previous_queue_item)
+         {
+            _free_operation_queue_item(queue_item);
```
Comment (davidgfnet) [inline cloud-storage/cloud_storage.c:787]: If you delete this item, the next iteration will execute `queue_item = queue_item->next`. But you just deleted it right? Sure the memory bits might still be there, but this is called "use after free" and it's not just dangerous (since malloc is free to reuse the memory inbetween the free and use) but also a classic and well known exploitation technique :P  See comment below on how to simplify this.

File: cloud-storage/providers/local_folder/list_files.c [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+   metadata->name = strdup(filename);
+   metadata->item_type = CLOUD_STORAGE_FILE;
+   metadata->type_data.file.hash_type = CLOUD_STORAGE_HASH_MD5;
+   metadata->type_data.file.hash_value = cloud_storage_get_md5_hash(absolute_path);
+   metadata->last_sync_time = time(NULL);
+
+   return metadata;
+}
+
+void cloud_storage_local_folder_list_files(cloud_storage_item_t *folder)
```
Comment (davidgfnet) [inline cloud-storage/providers/local_folder/list_files.c:51]: I'd love a couple comments here for a better understanding please

**Comment by njordan64** [comment]
Added comments and updated to use the new linked list and generic queue functions. Not tested yet, review should wait.

**Comment by njordan64** [comment]
Test this change. It it ready for review again.

File: cloud-storage/cloud_storage.c [hunk 13]
```diff
@@ -0,0 +1,947 @@
...
+static slock_t *_operation_queue_mutex;
+
+static scond_t *_operation_condition;
+static sthread_t *_operation_thread;
+
+static void _sync_files_execute(folder_type_t folder_type);
+
+static void _upload_file_execute(folder_type_t folder_type, char *file_name);
+
+void _cloud_storage_item_free_fn(void *item)
```
Comment (davidgfnet) [inline cloud-storage/cloud_storage.c:94]: If the goal is to make this function private, please add "static" so that it remains private within the compilation unit.

File: cloud-storage/cloud_storage.c [hunk 14]
```diff
@@ -0,0 +1,947 @@
...
+      slock_lock(_operation_queue_mutex);
+      if (generic_queue_length(_operation_queue) == 0)
+      {
+         scond_wait(_operation_condition, _operation_queue_mutex);
+      }
+
+      if (_shutdown)
+      {
+         slock_unlock(_operation_queue_mutex);
+         continue;
```
Comment (davidgfnet) [inline cloud-storage/cloud_storage.c:176]: Why continue? I mean it will indeed re-evaluate the while condition and stop iterating, but break is more readable right? Makes it clearer that you are aborting the loop.

File: cloud-storage/cloud_storage.c [hunk 15]
```diff
@@ -0,0 +1,947 @@
...
+            case DOWNLOAD_FILE:
+               break;
+         }
+
+         _free_operation(operation);
+      }
+   }
+
+   /* Broke out of the operation loop, will shutdown the operation thread */
+   if (_operation_queue)
```
Comment (davidgfnet) [inline cloud-storage/cloud_storage.c:201]: Wont this always be true? You create the thread after initializing the queue.

File: cloud-storage/cloud_storage.c [hunk 16]
```diff
@@ -0,0 +1,947 @@
...
+      return;
+   }
+
+   /* Signal the operation thread to shutdown in a critical section */
+   slock_lock(_operation_queue_mutex);
+   _shutdown = true;
+   scond_signal(_operation_condition);
+   slock_unlock(_operation_queue_mutex);
+
+   sthread_join(_operation_thread);
```
Comment (davidgfnet) [inline cloud-storage/cloud_storage.c:244]: I'd move the generic_queue_free(_operation_queue...) right after here. It's just clearer to have all the init/deinit stuff in the init/deinit functions right?

File: cloud-storage/cloud_storage.c (hunk 8 above)
Comment (davidgfnet) [reply cloud-storage/cloud_storage.c:273]: Yeah please, do not use calloc for a 16 byte buffer :D

File: cloud-storage/cloud_storage.c [hunk 17]
```diff
@@ -0,0 +1,947 @@
...
+}
+
+char *cloud_storage_get_md5_hash(char *absolute_filename)
```
Comment (davidgfnet) [inline cloud-storage/cloud_storage.c:247]: Could you please move this function to the `libretro-common/hash/lrc_hash.c` file? Or somewhere in common. There's already a function that does the same for SHA1 hashes (returns a hash for a given file path).

File: cloud-storage/cloud_storage.c [hunk 18]
```diff
@@ -0,0 +1,947 @@
...
+
+   /* Convert the MD5 checksum to a hex string */
+   checksum_str = bytes_to_hex_str(checksum, 16);
+   free(checksum);
+
+   return checksum_str;
+}
+
+/* Converts a value (game save/state directory setting) to an absolute directory path */
+static char *_config_value_to_directory(const char *value)
```
Comment (davidgfnet) [inline cloud-storage/cloud_storage.c:284]: Could it be that the "fill_pathname_parent_dir_name" function does the same as this function?
IIUC you are getting the basedir of a file/dir without the trailing slash right?

File: cloud-storage/providers/local_folder/upload_file.c [hunk 19]
```diff
@@ -0,0 +1,73 @@
...
+   dest_file = filestream_open(dest_filename, RETRO_VFS_FILE_ACCESS_WRITE, RETRO_VFS_FILE_ACCESS_HINT_NONE);
+   if (!dest_file) {
+      free(dest_filename);
+      filestream_close(src_file);
+      return false;
+   }
+   free(dest_filename);
+
+   /* Copy the file contents over 8k at a time */
+   bytes_read = filestream_read(src_file, buffer, BUFFER_SIZE);
```
Comment (davidgfnet) [inline cloud-storage/providers/local_folder/upload_file.c:63]: You can use do {} .. while (bytes_read > 0) perhaps if you think it's clearer.

File: cloud-storage/providers/local_folder/local_folder.c [hunk 20]
```diff
@@ -0,0 +1,104 @@
+/* Copyright  (C) 2020 The RetroArch team
+ *
+ * ---------------------------------------------------------------------------------------
+ * The following license statement only applies to this file (google.c).
```
Comment (davidgfnet) [inline cloud-storage/providers/local_folder/local_folder.c:4]: google.c?

File: cloud-storage/providers/local_folder/local_folder_internal.h [hunk 21]
```diff
@@ -0,0 +1,65 @@
...
+ */
+
+#include <retro_common_api.h>
+
+#include <formats/rjson.h>
+#include <net/net_http.h>
+
+#include "../../cloud_storage.h"
+
+#ifndef _CLOUD_STORAGE_GOOGLE_INTERNAL_H
```
Comment (davidgfnet) [inline cloud-storage/providers/local_folder/local_folder_internal.h:30]: Copy-paste :)

File: cloud-storage/provider_common.h [hunk 22]
```diff
@@ -0,0 +1,44 @@
...
+ * @brief Get the hex encoded MD5 hash of a local file
+ *
+ * Calculates the MD5 checksum of the contents of a local file. The checksum
+ * is hex encoded and returned a newly allocated char*. Caller must release
+ * the returned string.
+ *
+ * @param absolute_filename absolute filename of the local file
+ * @return hex encoded MD5 checksum of file contents
+ */
+char *cloud_storage_get_md5_hash(char *absolute_filename);
```
Comment (davidgfnet) [inline cloud-storage/provider_common.h:40]: As mentioned before you can move this to common and you can get rid of this file then.

File: cloud-storage/cloud_storage.c (hunk 16 above)
Comment (njordan64) [reply cloud-storage/cloud_storage.c:233]: Changed.

File: cloud-storage/cloud_storage.c (hunk 9 above)
Comment (njordan64) [reply cloud-storage/cloud_storage.c:499]: I am not clear on what you are looking for here? I realize the code is far from acceptable quality, but there are now many comments. There is a comment two lines up and five lines down.

File: cloud-storage/cloud_storage.c (hunk 13 above)
Comment (njordan64) [reply cloud-storage/cloud_storage.c]: Changed to be static. Good catch.

File: cloud-storage/cloud_storage.c (hunk 14 above)
Comment (njordan64) [reply cloud-storage/cloud_storage.c]: Updated.

File: cloud-storage/cloud_storage.c (hunk 15 above)
Comment (njordan64) [reply cloud-storage/cloud_storage.c]: This was a mistake from changing to the common linked list. Fixed now.

File: cloud-storage/cloud_storage.c (hunk 17 above)
Comment (njordan64) [reply cloud-storage/cloud_storage.c]: Moved.

File: cloud-storage/cloud_storage.c (hunk 18 above)
Comment (njordan64) [reply cloud-storage/cloud_storage.c]: The function "fill_pathname_parent_dir_name" gets the name of the immediate parent directory. I want the absolute path minus the last path component.
This has been changed to use "fill_pathname_basedir".

File: cloud-storage/providers/local_folder/upload_file.c (hunk 19 above)
Comment (njordan64) [reply cloud-storage/providers/local_folder/upload_file.c]: Hard to say what is better. I changed to a do..while loop. Only need one line to read bytes, but had to introduce an if statement.

File: cloud-storage/providers/local_folder/local_folder.c (hunk 20 above)
Comment (njordan64) [reply cloud-storage/providers/local_folder/local_folder.c]: This file was removed.

File: cloud-storage/providers/local_folder/local_folder_internal.h (hunk 21 above)
Comment (njordan64) [reply cloud-storage/providers/local_folder/local_folder_internal.h]: This file was removed.

File: cloud-storage/provider_common.h (hunk 22 above)
Comment (njordan64) [reply cloud-storage/provider_common.h]: This was moved to libretro-common.

**Comment by njordan64** [comment]
@davidgfnet @twinaphex I'm not sure what is going on with the CI build error. My best guess is that it didn't checkout the correct branch for building.
The line with the error does not match up with the source in my branch.

File: cloud-storage/cloud_storage.c (hunk 9 above)
Comment (davidgfnet) [reply cloud-storage/cloud_storage.c:499]: Yeah previously there weren't any comments at all, that's why I added the comment. Thanks!
However it is still not clear to me why you do this conditionally. Why do you need to *sometimes* call basedir and some times not?
What kind of file path do you expect to receive here? I take you receive a file (xx.yyy) and you get the basedir to get the directory it's stored in right? Why not checking whether the path is a file, and if it is, take the base dir? Would it achieve the same? I just have the feeling (but I can be totally wrong! not pointing any fingers!) that this looks like a hack.

**Comment by Ghabry** [comment]
does this go through the VFS or how are cores supposed to opt-in into this feature?

File: cloud-storage/cloud_storage.c (hunk 9 above)
Comment (njordan64) [reply cloud-storage/cloud_storage.c:499]: Save state can have values such as
- /home/user/.local/retroarch/states/.state
- /home/user/.local/retroarch/states

In the first case, it is specifying a directory with an extension. In the second case it only specifies a directory. If it is the first case, I need to remove the extension. In the second case, I need to preserve the path as it is.

Far better would be to have separate settings for the save state directory and extension. I'm fine making that change, but it would be a separate PR and have concerns around backwards compatibility.

**Comment by njordan64** [comment]
@davidgfnet Is there anything remaing for this PR? I think I have addressed all open comments.

**Comment by njordan64** [comment]
> does this go through the VFS or how are cores supposed to opt-in into this feature?

This is not using VFS. There are many different VFS implementations and they vary by OS/desktop environment. Cores do not opt in. It is hooked in at a lower level. When RetroArch is saving a state, it could also be uploaded to a cloud account (just a different directory in this PR).

**Comment by njordan64** [comment]
This patch is unworkable.
